### ISSUE TITLE
Grid.py: global name 'time' is not defined

### ISSUE THREAD

**Issue by dggamez** [ganga-devs/ganga#111]
When I submit a job (with backend=CREAM()) using the cvmfs latest version (or many others) of ganga I get the error shown below. It does not happen when using a quite old version like 6-0-42. I would appreciate any help on this.

Thanks!
```
Ganga.GPIDev.Lib.Job               : INFO     submitting job 55
Ganga.GPIDev.Lib.Job               : INFO     job 55 status changed to "submitting"
Ganga.Lib.Executable               : INFO     Preparing Executable application.
Ganga.Lib.Executable               : INFO     Created shared directory: conf-d43e54a8-6166-40c6-9a6e-4c1164c08bcc
Ganga.GPIDev.Lib.Job               : INFO     Preparing subjobs
Ganga.Lib.LCG                      : WARNING  CREAM CE assigment from AtlasCREAMRequirements failed.
Ganga.GPIDev.Lib.Job               : WARNING  ---------- error in user/extension code ----------
Ganga.GPIDev.Lib.Job               : WARNING  Traceback (most recent call last):
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.6-hotfix1/python/Ganga/GPIDev/Lib/Job/Job.py", line 1592, in submit
Ganga.GPIDev.Lib.Job               : WARNING      r = self.backend.master_submit( rjobs, jobsubconfig, jobmasterconfig)
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.6-hotfix1/python/Ganga/Lib/LCG/CREAM.py", line 1086, in master_submit
Ganga.GPIDev.Lib.Job               : WARNING      if not grids['GLITE'].cream_proxy_delegation(self.CE):
Ganga.GPIDev.Lib.Job               : WARNING    File "/cvmfs/ganga.cern.ch/Ganga/install/6.1.6-hotfix1/python/Ganga/Lib/LCG/Grid.py", line 914, in cream_proxy_delegation
Ganga.GPIDev.Lib.Job               : WARNING      t_expire = time.time() + \
Ganga.GPIDev.Lib.Job               : WARNING  NameError: global name 'time' is not defined
Ganga.GPIDev.Lib.Job               : WARNING  --------------------------------------------------
Ganga.GPIDev.Lib.Job               : ERROR    global name 'time' is not defined ... reverting job 55 to the new status
Ganga.GPIDev.Lib.Job               : INFO     job 55 status changed to "new"
Ganga.Runtime.bootstrap            : ERROR    JobError: global name 'time' is not defined
```

**Comment by milliams** [comment]
You are running 6.1.6-hotfix1 which is not the latest version. 6.1.14 is the latest version on CVMFS which can be run via `/cvmfs/ganga.cern.ch/Ganga/install/LATEST/bin/ganga`. Do you still see this bug in the latest version?

**Comment by drmarkwslater** [comment]
Indeed - this is because at the workshop yesterday I recommended people use 6.1.6-hotfix1 from CVMFS to avoid the on-going syncing issues with the repo. However, it looks like I fixed this in a previous issue #55. 

Having said that , it would be good to check that the CREAM backend is still working OK in the latest release!

**Comment by rob-c** [comment]
@drmarkwslater Is this a bug with our repo syncing or some external repo to Core Ganga?

**Comment by drmarkwslater** [comment]
@rob-c I'm afraid it's still a syncing bug in Core. I hadn't seen it on develop branch with the Local backend but if you run a Dirac job, the job submits fine but ~50% of the time, (possibly if you don't restart Ganga before hand and it's an Exe+Dirac job) when the job completes, the completing messages go through with an additional warning of 'User changed status from submitted -> running' or something but you do 'jobs' and it still reports submitted. Restart ganga and it's all OK and the job completes again properly.

As I say, I hadn't seen this on the Local backend but certainly a few users running off CVMFS on CernVM saw a similar thing for Local Exe jobs, usually involving job copies, but the job would submit, complete and then if you do 'jobs' it still reports as new. 

As soon as I have time I'll right up the bugs that were found yesterday as issues!

**Comment by rob-c** [comment]
Strange, I don't see this with GangaLHCb and GangaDirac but then I think the executable has been changed in this workflow...

If this is a problem with migrating the status forward in:

submitting->submitted->running->completing->complete(failed) then I don't know what could be causing the problems here

If the problem is related to jumping back from the 'completing' state to the submitted state then I suspect that the Dirac.reset() code is being called or the job is failing in such a way that the status is being reset to try and recover the failing job.

If the 'jobs' is becoming out of sync with the real job status this is likely due to the horrible way that we pass around the ```_index_cache``` without regenerating it upon request which avoids these problems and is a relatively lightweight function.

I'm seeing separate problems that some job copies aren't being picked up by the monitoring and from looking at the code there the object remains unknown to the monitoring #85 as far as I can tell as I see no warning/errors.

I panicked hearing repo syncing as I was expecting a bug relating to registry/repo transactions failing or throwing exceptions.

**Comment by milliams** [comment]
The main bug we were seeing in the session yesterday seemed to be people creating simple Local/Executable job, submitting them and then somehow they were ending up back in the `new` state. Trying to submit them as this point gave a "cannot do state transition"-type error and the job would not work after that. It felt like the registry was doing a load at the wrong time and overwriting the local state (similar to what I'd seen with jobtree previously) but I'm not sure.

I've tried to reproduce this locally but have been unable. The users were using CERN VM with the GridPP context and running 6.1.14 from CVMFS.

**Comment by rob-c** [comment]
As for the job transitions I've fixed this as part of #117 I don't know about the original problem to do with time being out of scope.

There is an ```import time``` at the top level of the file in develop, so if this was there in earlier versions it wasn't exposed in the local scope of this function for some reason.

**Comment by milliams** [comment]
This was fixed back in 8d0fc46c06ab30ea404d3d4dee6e9ca0d6bca591.
